client-apis/encrypt.md

```diff
@@ -113,68 +171,185 @@ This document refers to the output of the key derivation algorithm as the derive
-  if included.
-- [IV](../data-format/message-header.md#iv): MUST have a value of 0, padded to the [IV length](../data-format/message-header.md#iv-length).
-- [Authentication Tag](../data-format/message-header.md#authentication-tag): MUST be the output of the
-  [authenticated encryption algorithm](../framework/algorithm-suites.md#encryption-algorithm)
-  specified by the [algorithm suite](../framework/algorithm-suites.md), with the following inputs:
-  - The AAD is the serialized [message header body](../data-format/message-header.md#header-body)
-  - The IV is the [IV specified above](../data-format/message-header.md#iv)
-  - The cipherkey is the derived data key
-  - The plaintext is an empty byte array
+- [Frame Length](../data-format/message-header.md#frame-length): MUST be the value of the frame size determined above.
+
+After serializing the message header body,
+this operation MUST calculate an [authentication tag](../data-format/message-header.md#authentication-tag)
+over the message header body.
+The value of this MUST be the output of the [authenticated encryption algorithm](../framework/algorithm-suites.md#encryption-algorithm)
+specified by the [algorithm suite](../framework/algorithm-suites.md), with the following inputs:
+
+- The AAD is the serialized [message header body](../data-format/message-header.md#header-body).
+- The IV has a value of 0.
+- The cipherkey is the derived data key
+- The plaintext is an empty byte array
+
+With the authentication tag calculated, this operation MUST serialize the
+[message header authentication](../data-format/message-header.md#header-authentication)
+with the following specifics.
+
+- [IV](../data-format/message-header.md#iv): MUST have a value of 0,
+  padded to the [IV length](../data-format/message-header.md#iv-length).
+- [Authentication Tag](../data-format/message-header.md#authentication-tag): MUST have the value
+  of the authentication tag calculated above.
+
+Once the entire message header has been authenticated and serialized,
+the serialized message header MAY be released.
+
+The encrypted message outputted by this operation MUST have a message header equal
+to the message header calculated in this step.
+
+If the algorithm suite contains a signature algorithm and
+this operation is streaming the encrypted message output to the caller,
+this operation MUST input the serialized header to the signature algorithm as soon as it is serialized,
+such that the serialized header isn't required to remain in memory to complete
+the [signature calculation](#signature-calculation).
+
+## Construct the body
+
+The encrypted message outputted by this operation MUST have a message body equal
+to the message body calculated in this step.
+
+While there MAY still be plaintext left to encrypt,
+this operation MUST either wait for more plaintext to become available,
+wait for an end to the plaintext to be indicated,
+or to perform encryption on the available plaintext to construct frames
```

Comment:
Should `or to perform encryption` be `or perform encryption`? I don't think it's a `or wait to perform encryption`, but I could be wrong.